File: tests/testprovider/random.go
// Copyright 2016-2021, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//go:build !all
// +build !all

package main

import (
	"context"
	"crypto/rand"
	"fmt"
	"math/big"

	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
	rpc "github.com/pulumi/pulumi/sdk/v3/proto/go"

	"google.golang.org/protobuf/types/known/emptypb"
)

func init() {
	providerSchema.Resources["testprovider:index:Random"] = pschema.ResourceSpec{
		ObjectTypeSpec: pschema.ObjectTypeSpec{
			Description: "A test resource that generates a random string of a given length and with an optional prefix.",
			Properties: map[string]pschema.PropertySpec{
				"length": {
					TypeSpec:    pschema.TypeSpec{Type: "integer"},
					Description: "The length of the random string (not including the prefix, if any).",
				},
				"prefix": {
					TypeSpec:    pschema.TypeSpec{Type: "string"},
					Description: "An optional prefix.",
				},
				"result": {
					TypeSpec:    pschema.TypeSpec{Type: "string"},
					Description: "A random string.",
				},
			},
			Type: "object",
		},
		InputProperties: map[string]pschema.PropertySpec{
			"length": {
				TypeSpec:    pschema.TypeSpec{Type: "integer"},
				Description: "The length of the random string (not including the prefix, if any).",
			},
			"prefix": {
				TypeSpec:    pschema.TypeSpec{Type: "string"},
				Description: "An optional prefix.",
			},
		},
	}
}

type randomProvider struct{}

func (p *randomProvider) Check(ctx context.Context, req *rpc.CheckRequest) (*rpc.CheckResponse, error) {
	return &rpc.CheckResponse{Inputs: req.News, Failures: nil}, nil
}

func (p *randomProvider) Diff(ctx context.Context, req *rpc.DiffRequest) (*rpc.DiffResponse, error) {
	olds, err := plugin.UnmarshalProperties(req.GetOlds(), plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true})
	if err != nil {
		return nil, err
	}

	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true})
	if err != nil {
		return nil, err
	}

	d := olds.Diff(news)
	var replaces []string
	changes := rpc.DiffResponse_DIFF_NONE
	if d.Changed("length") {
		changes = rpc.DiffResponse_DIFF_SOME
		replaces = append(replaces, "length")
	}

	if d.Changed("prefix") {
		changes = rpc.DiffResponse_DIFF_SOME
		replaces = append(replaces, "prefix")
	}

	return &rpc.DiffResponse{
		Changes:  changes,
		Replaces: replaces,
	}, nil
}

func (p *randomProvider) Create(ctx context.Context, req *rpc.CreateRequest) (*rpc.CreateResponse, error) {
	inputs, err := plugin.UnmarshalProperties(req.GetProperties(), plugin.MarshalOptions{
		KeepUnknowns: true,
		SkipNulls:    true,
	})
	if err != nil {
		return nil, err
	}

	if !inputs["length"].IsNumber() {
		return nil, fmt.Errorf("expected input property 'length' of type 'number' but got '%s", inputs["length"].TypeString())
	}

	n := int(inputs["length"].NumberValue())

	var prefix string
	if p, has := inputs["prefix"]; has {
		if !p.IsString() {
			return nil, fmt.Errorf("expected input property 'prefix' of type 'string' but got '%s", p.TypeString())
		}
		prefix = p.StringValue()
	}

	// Actually "create" the random number
	result, err := makeRandom(n)
	if err != nil {
		return nil, err
	}

	outputs := resource.NewPropertyMapFromMap(map[string]interface{}{
		"length": n,
		"result": prefix + result,
	})
	if prefix != "" {
		outputs["prefix"] = resource.NewStringProperty(prefix)
	}
	outputs["result"] = resource.MakeSecret(outputs["result"])

	outputProperties, err := plugin.MarshalProperties(
		outputs,
		plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true},
	)
	if err != nil {
		return nil, err
	}
	return &rpc.CreateResponse{
		Id:         result,
		Properties: outputProperties,
	}, nil
}

func (p *randomProvider) Read(ctx context.Context, req *rpc.ReadRequest) (*rpc.ReadResponse, error) {
	// Just return back the input state.
	return &rpc.ReadResponse{
		Id:         req.Id,
		Properties: req.Properties,
	}, nil
}

func (p *randomProvider) Update(ctx context.Context, req *rpc.UpdateRequest) (*rpc.UpdateResponse, error) {
	// Our Random resource will never be updated - if there is a diff, it will be a replacement.
	panic("Update not implemented")
}

func (p *randomProvider) Delete(ctx context.Context, req *rpc.DeleteRequest) (*emptypb.Empty, error) {
	// Note that for our Random resource, we don't have to do anything on Delete.
	return &emptypb.Empty{}, nil
}

func (p *randomProvider) Invoke(ctx context.Context, req *rpc.InvokeRequest) (*rpc.InvokeResponse, error) {
	// The random provider doesn't support any invokes currently.
	panic("Invoke not implemented")
}

func (p *randomProvider) Call(ctx context.Context, req *rpc.CallRequest) (*rpc.CallResponse, error) {
	// The random provider doesn't support any call currently.
	panic("Call not implemented")
}

func makeRandom(length int) (string, error) {
	charset := []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
	result := make([]rune, length)
	for i := range result {
		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(charset))))
		if err != nil {
			return "", err
		}
		result[i] = charset[num.Int64()]
	}
	return string(result), nil
}

File: tests/testprovider-py/src/random.py
import random
import string
from pulumi.provider.experimental.provider import DiffRequest, Provider, CreateResponse, DiffResponse, CheckResponse, ReadResponse
from pulumi.provider.experimental.property_value import PropertyValue

class RandomProvider(Provider):
    """
    A provider for a resource that generates a random string of a given length with an optional prefix.
    """

    schema = {
        "resources": {
            "testprovider:index:Random": {
                "description": "A test resource that generates a random string of a given length and with an optional prefix.",
                "type": "object",
                "properties": {
                    "length": {
                        "type": "integer",
                        "description": "The length of the random string (not including the prefix, if any).",
                    },
                    "prefix": {
                        "type": "string",
                        "description": "An optional prefix.",
                    },
                    "result": {
                        "type": "string",
                        "description": "A random string.",
                    },
                },
                "inputProperties": {
                    "length": {
                        "type": "integer",
                        "description": "The length of the random string (not including the prefix, if any).",
                    },
                    "prefix": {
                        "type": "string",
                        "description": "An optional prefix.",
                    },
                },
            }
        }
    }

    async def check(self, request):
        inputs = request.new_inputs
        if "length" not in inputs:
            raise ValueError("Missing required property 'length'")
        if "prefix" not in inputs:
            inputs["prefix"] = PropertyValue("")

        length = inputs.get("length").value
        if not isinstance(length, float):
            raise ValueError(f"Expected 'length' to be a number, got {type(length)}")

        prefix = inputs.get("prefix").value
        if not isinstance(prefix, str):
            raise ValueError(f"Expected 'prefix' to be a string, got {type(prefix)}")

        return CheckResponse(inputs=inputs)

    async def diff(self, request: DiffRequest):
        replaces = []
        if request.old_inputs.get("length") != request.new_inputs.get("length"):
            replaces.append("length")
        if request.old_inputs.get("prefix") != request.new_inputs.get("prefix"):
            replaces.append("prefix")

        return DiffResponse(
            changes=bool(replaces),
            replaces=replaces,
        )

    async def create(self, request):
        length = request.properties.get("length").value
        prefix = request.properties.get("prefix").value

        result = self._make_random(int(length))
        full_result = f"{prefix}{result}"

        return CreateResponse(
            resource_id=full_result,
            properties={
                "length": PropertyValue(length),
                "prefix": PropertyValue(prefix),
                "result": PropertyValue(full_result),
            },
        )

    async def delete(self, request):
        pass

    @staticmethod
    def _make_random(length: int) -> str:
        """
        Generates a random string of the given length.
        """
        charset = string.ascii_letters + string.digits
        return ''.join(random.choice(charset) for _ in range(length))